File: code_snippets/ts/src/concepts/services/subscription_service.ts
import * as restate from "@restatedev/restate-sdk";

// <start_here>
// <mark_2>
const subscriptionService = restate.service({
  // </mark_2>
  name: "SubscriptionService",
  // <mark_2>
  handlers: {
    add: async (ctx: restate.Context, req: SubscriptionRequest) => {
      // </mark_2>
      // <mark_3>
      // <mark_1>
      const paymentId = ctx.rand.uuidv4();
      // </mark_1>

      // <mark_1>
      const payRef = await ctx.run(() =>
        createRecurringPayment(req.creditCard, paymentId)
      );
      // </mark_1>

      for (const subscription of req.subscriptions) {
        // <mark_1>
        await ctx.run(() =>
          createSubscription(req.userId, subscription, payRef)
        );
        // </mark_1>
      }
      // </mark_3>
    },
  },
});

// <mark_4>
restate.endpoint().bind(subscriptionService).listen(9080);
// </mark_4>
// <end_here>

type SubscriptionRequest = {
  creditCard: string;
  userId: any;
  subscriptions: any[];
};

function createRecurringPayment(creditCard: string, paymentId: string) {}

function createSubscription(userId: any, subscription: any, payRef: void) {
  return undefined;
}

File: code_snippets/go/concepts/services/subscriptionservice.go
package main

import (
	"context"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log"
)

type SubscriptionRequest struct {
	UserID        string   `json:"userId"`
	CreditCard    string   `json:"creditCard"`
	Subscriptions []string `json:"subscriptions"`
}

// <start_here>
// <mark_2>
type SubscriptionService struct{}

// </mark_2>

// <mark_2>
func (SubscriptionService) Add(ctx restate.Context, req SubscriptionRequest) error {
	// </mark_2>
	// <mark_3>
	// <mark_1>
	paymentId := restate.Rand(ctx).UUID().String()
	// </mark_1>

	// <mark_1>
	payRef, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return CreateRecurringPayment(req.CreditCard, paymentId)
	})
	// </mark_1>
	if err != nil {
		return err
	}

	for _, subscription := range req.Subscriptions {
		// <mark_1>
		if _, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
			return restate.Void{}, CreateSubscription(req.UserID, subscription, payRef)
			// </mark_1>
		}); err != nil {
			return err
		}
	}

	return nil
	// </mark_3>
}

func main() {
	// <mark_4>
	if err := server.NewRestate().
		Bind(restate.Reflect(SubscriptionService{})).
		Start(context.Background(), ":9080"); err != nil {
		log.Fatal(err)
	}
	// </mark_4>
}

// <end_here>

func CreateSubscription(id string, subscription string, ref string) error {
	return nil
}

func CreateRecurringPayment(card string, id string) (string, error) {
	return id, nil
}
